### ISSUE TITLE
How to detect if PBR is enabled

### ISSUE THREAD

**Issue by sacereda** [assimp/assimp#4424]
After updating to v5.2.2, I'm struggling on how to detect if a material is PBR or regular. Based on this commit from Obj importer:

https://github.com/assimp/assimp/blob/bbcefbd0348c5407e57f9f66a16a69b61159f537/code/AssetLib/Obj/ObjFileImporter.cpp#L620

https://github.com/assimp/assimp/commit/19371af6e65608ffc968df646bf20278e6d99414

It seems PBR properties like AI_MATKEY_METALLIC_FACTOR or AI_MATKEY_ROUGHNESS_FACTOR are added always, so I cannot check if the material has any of those in order to apply or not PBR properties on my application.

I was thinking maybe material properties like AI_MATKEY_USE_METALLIC_MAP or AI_MATKEY_USE_ROUGHNESS_MAP, but they are only added in FBXConverter.

Maybe there should be an AI_MATKEY_USE_PBR property or similar?

**Comment by TerenceRussell** [comment]
I'm not sure about the general case, but for OBJ models a PBR hint flag could be added to the material.

If any of the newer PRB fields are encountered on load, the flag would be set.

The hint could be retrieved from the material like so:

`int isMaterialPBR = 0; // 0 = no; 1 = possibly.`
`aiGetMaterialInteger(mtl, AI_MATKEY_OBJ_PBR_HINT, &isMaterialPBR);`


This would be simple to add.

**Comment by sacereda** [comment]
Hi Terence,

I think this will hide the root problem, as the obj is still exporting PBR properties on non-pbr materials, and we will end up adding an importer-specific PBR hints on each importer.

Please see the output from 'assimp info' (after applying https://github.com/assimp/assimp/pull/4426):

> $ bin\debug\assimpd.exe info test\models\OBJ\box.obj
...
Named Materials:
    'Default' (prop) [index / bytes | texture semantic]
        0 (?mat.name): [0 / 12 | n/a]
        1 ($mat.shadingm): [0 / 4 | n/a]
        2 ($mat.illum): [0 / 4 | n/a]
        3 ($clr.ambient): [0 / 12 | n/a]
        4 ($clr.diffuse): [0 / 12 | n/a]
        5 ($clr.specular): [0 / 12 | n/a]
        6 ($clr.emissive): [0 / 12 | n/a]
        7 ($mat.shininess): [0 / 4 | n/a]
        8 ($mat.opacity): [0 / 4 | n/a]
        9 ($clr.transparent): [0 / 12 | n/a]
        10 ($mat.roughnessFactor): [0 / 4 | n/a]
        11 ($mat.metallicFactor): [0 / 4 | n/a]
        12 ($clr.sheen.factor): [0 / 12 | n/a]
        13 ($mat.clearcoat.factor): [0 / 4 | n/a]
        14 ($mat.clearcoat.roughnessFactor): [0 / 4 | n/a]
        15 ($mat.anisotropyFactor): [0 / 4 | n/a]
        16 ($mat.refracti): [0 / 4 | n/a]
...

I can try to send another PR on the OBJ importer, just wanted to ask if I was missing or doing something wrong when checking for PBR.

**Comment by sacereda** [comment]
Hi Terence,

Thanks for the suggestion, I'll do that.

Regards

**Comment by sacereda** [comment]
@errissa,

After https://github.com/assimp/assimp/pull/4272, any obj imported with assimp will have materials with metallicFactor, roughnessFactor, sheenFactor, etc, although they are not present in the .mtl (see https://github.com/assimp/assimp/issues/4424#issuecomment-1061568962).

I have a local branch with some changes regarding that. Basically, those fields from ObjFileData::Material are optional (using a class similar to STEPFile.h::Maybe), and they are not added as properties unless they are found in the mtl.

For example, with metallicFactor:

```
-    ai_real roughness;
+    Maybe<ai_real> roughness;

...

-            metallic(ai_real(0.0)),
+            metallic(),

...

-        mat->AddProperty(&pCurrentMaterial->metallic, 1, AI_MATKEY_METALLIC_FACTOR);
+        if (pCurrentMaterial->metallic)
+            mat->AddProperty(&pCurrentMaterial->metallic.Get(), 1, AI_MATKEY_METALLIC_FACTOR);

...

             case 'm':
                 ++m_DataIt;
-                getFloatValue(m_pModel->m_pCurrentMaterial->metallic);
+                m_pModel->m_pCurrentMaterial->metallic = getMaybeFloatValue();
                 break;
```

What do you think about something like this?

Regards,

**Comment by errissa** [comment]
@sacereda Your approach seems reasonable to me. Currently if the PBR property is missing from the mtl a default reasonable for a PBR workflow gets set but as user you can’t tell if it’s a default because it was missing or was the model author’s intention.

I do think it would be nice to have a higher level flag or set of flags for the intended workflow. Not just PBR (metallic/roughness) but also the various KHR extensions for glTF such as KHR_materials_pbrSpecularGlossiness, KHR_materials_transmission. But this is separate from the OBJ issue.

**Comment by sacereda** [comment]
@errissa Yes, a default value for an specific property could be good, but I normally do it in the app itself, so if the property is not found I can use a default value if needed. So this way you can choose to apply it or not.

Thanks for the feedback, I'll push a PR for the obj issue and ping you there.

**Comment by RichardTea** [comment]
`AI_MATKEY_SHADING_MODEL` `aiShadingMode_PBR_BRDF` is intended to be the flag for "this is genuinely PBR"

That said, I'm not certain whether applications can rely on this at the moment.
My application only consumes PBR, so I simply assume importers that create any PBR properties did a better job than my fallback conversion and never look for the aiShadingMode_PBR_BRDF shading model.

**Comment by tellypresence** [comment]
@sacereda Was this fixed by #4440?

**Comment by sacereda** [comment]
Yes,  I forgot to update this one. Thanks.
